addons/hr_holidays/models/hr_leave.py

```diff
@@ -107,45 +107,133 @@ def _default_employee(self):
     can_reset = fields.Boolean('Can reset', compute='_compute_can_reset')
     can_approve = fields.Boolean('Can Approve', compute='_compute_can_approve')
 
+    # Those fields are mostly for the interface only
+
+    # Interface fields used when not using hour-based computation
+    request_date_from = fields.Date('Request Start Date')
+    request_date_to = fields.Date('Request End Date')
+
+    leave_type_request_unit = fields.Selection(related='holiday_status_id.request_unit', readonly=True)
+
+    # These fields are only used only when the leave is taken in half days
+    request_date_from_period = fields.Selection([('am', 'Morning'),
+                                                 ('pm', 'Afternoon')],
+                                                string="Date Period Start", default='am')
+    request_date_to_period = fields.Selection([('am', 'Morning'),
+                                               ('pm', 'Afternoon')],
+                                              string="Date Period End", default='pm')
+
+    request_unit_all = fields.Selection([('half', 'Half Day'),
+                             ('day', '1 Day'),
+                             ('period', 'Period')],
+                            default=lambda req: 'period' if req.leave_type_request_unit == 'hour' else 'half')
+    # Duplicate field because we cannot hide some entries of a selection field
+    request_unit_day = fields.Selection([('day', '1 Day'),
+                                 ('period', 'Period')],
+                                default='day')
+
+    number_of_hours = fields.Float(
+        'Hours Allocation', copy=False, readonly=True, compute='_compute_number_of_hours',
+        help='Number of hours of the leave request according to your working schedule.')
+
     _sql_constraints = [
         ('type_value', "CHECK( (holiday_type='employee' AND employee_id IS NOT NULL) or (holiday_type='category' AND category_id IS NOT NULL) or (holiday_type='department' AND department_id IS NOT NULL) )",
          "The employee, department or employee category of this request is missing. Please make sure that your user login is linked to an employee."),
         ('date_check2', "CHECK ((date_from <= date_to))", "The start date must be anterior to the end date."),
-        ('date_check', "CHECK ( number_of_days_temp >= 0 )", "The number of days must be greater than 0."),
+        ('date_check', "CHECK ( number_of_days_temp >= 0 )", "If you want to change the number of days you should use the 'period' mode"),
     ]
 
+    @api.onchange('request_unit_all', 'request_date_from_period', 'request_date_to_period',
+                  'holiday_status_id', 'request_date_from', 'request_date_to')
+    def _onchange_request_parameters(self):
+        if self.request_date_from:
+            if self.date_from:
```

Comment:
Hello,

The method seems already a lot better! Don't you think employee_id should be in the onchange dependencies as you depend on its calendar ?

There is still some confusion about date_(from/to) and self.date_(from/to). INdeed self.date_from and date_from are computed / rewritten several times in the method. It is therefore not easy to follow the path and see hen the update is done.

I would write on self.date_from and date_to only at the end of the method and use variables during the computation. Maybe we can have an exception with the preliminary return if we don't have both request_date_from and request_date_to, in which case it should be in the "if" .